s2/edge_query_test.go

```diff
@@ -177,6 +177,103 @@ func TestEdgeQuerySortAndUnique(t *testing.T) {
 	}
 }
 
+func TestEdgeQueryOptimized(t *testing.T) {
+	tests := []struct {
+		name string
+		locs []struct{ latitude, longitude float64 }
+	}{
+		{
+			// 0 intermediate faces: first and last handled directly
+			"faces 0,4",
+			[]struct{ latitude, longitude float64 }{{20, 20}, {40, -100}},
+		},
+
+		{
+			"faces 0,4 (3+1 shapes)",
+			[]struct{ latitude, longitude float64 }{{20, 20}, {25, 25}, {15, 15}, {40, -100}},
+		},
+
+		{
+			// 1 intermediate face with 1 shape each: addInitialRange covers adequately
+			"faces 0,1,4",
+			[]struct{ latitude, longitude float64 }{{20, 20}, {40, 90}, {40, -100}},
+		},
+
+		{
+			// 1 intermediate face but multiple shapes on first face: first top-level cell
+			// doesn't cover all shapes, and addInitialRange misses earlier ones
```

Comment:
Don't refer to the buggy state in the comments.  Those comments are essentially already out of date.